bundles/alpha.model/src/alpha/model/util/AffineFactorizer.xtend

```diff
@@ -0,0 +1,391 @@
+	 *         from each of the input expressions to the "remaining term" that is left after
+	 *         pulling out the common factor. That is, the following would return true,
+	 *         where <code>expr</code> is one of the input expressions and <code>result</code>
+	 *         is the return value of this function:
+	 *         <code>expr.isPlainEqual(result.value.get(expr).pullback(result.key.copy))</code>
+	 */
+	def static factorizeExpressions(ISLMultiAff... expressions) {
+		// Name all the outputs from the expressions so we can track them.
+		// Note: this is saved as a map from the original expression to the named one.
+		val named = expressions.nameExpressionOutputs
+		
+		// Merge everything into one big expression, and decompose it.
+		// Since we need them again later, copy the named expressions first.
+		val decomposed = named.values.map[it.copy].mergeExpressions.decomposeExpression
+		val hExpression = decomposed.key
+		val qExpression = decomposed.value
+		
+		// Construct a map from each of the original expressions
+		// to the projection of Q which is associated with it.
+		// Wrap it in a HashMap to avoid recomputing values whenever they're accessed.
+		val projectionMap = new HashMap<ISLMultiAff, ISLMultiAff>(
+			expressions.toInvertedMap[expr | getDecompositionProjection(named.get(expr), qExpression)])
+		return hExpression -> projectionMap
+	}
+
+
+	////////////////////////////////////////////////////////////
+	// Merging Expressions Into One
+	////////////////////////////////////////////////////////////
+
+	/** 
+	 * Creates a copy of each input expression, but renames all the outputs so they can be tracked.
+	 * Returns a map from the original expressions to the named ones.
+	 */
+	def static nameExpressionOutputs(ISLMultiAff... expressions) {
+		val names = getNameGenerator("orig_out")
+		return expressions
+			.toInvertedMap[nameSingleExpressionOutputs(it, names)]
+			.toHashMap
+	}
+
+	/** Generates names containing the given prefix and an incrementing value, separated by an underscore. */
+	def private static getNameGenerator(String prefix) {
+		return (0..Integer.MAX_VALUE)
+			.map[prefix + "_" + it.toString()]
+			.iterator()
+	}
+
+	/** Gives each output of the provided expression a unique name, overwriting any existing names. */
+	def private static nameSingleExpressionOutputs(ISLMultiAff expression, Iterator<String> names) {
+		val outCount = expression.dim(ISLDimType.isl_dim_out)
+		val assignName = [ISLMultiAff expr, int outIndex | expr.setDimName(ISLDimType.isl_dim_out, outIndex, names.next)]
+		val merged = (0 ..< outCount).fold(expression.copy, assignName)
+		return merged
+	}
+	
+	/**
+	 * Wraps a map into a hash map so its values are only computed once.
+	 * Intended for use with Xtend's iterable/map extensions, which can recompute values
+	 * each time they're accessed due to lazy evaluation.
+	 */
+	def private static <K,V> toHashMap(Map<K,V> map) {
```

Comment:
Is this really needed? I thought there were “built-in” function to xtend that do this already. If not, are there performance concerns?

Reply:
Xtend does have a built-in function which claims to do something similar. However, the class it returns will re-compute the values each time you try to access them, which results in many calls to ISL and duplicate objects being created. This caused me a lot of issues, but I found that wrapping it in a HashMap prevents the constant re-computation.

I also looked at putting this extension method somewhere more global so it can be used more easily, but I couldn't find a way to have Xtend import that function in a way that let me use it, so I just have it here.

Reply:
Gotcha, ok that works for me in this case.